README.md
tm-node-mpd
============

[![npm](https://img.shields.io/npm/v/tm-node-mpd.svg)](https://www.npmjs.org/package/tm-node-mpd)

This is fork of [node-mpd](https://github.com/Prior99/node-mpd) project by Frederick Gnodtke.

tm-node-mpd is a library for simple communicating with a [music player daemon](http://www.musicpd.org/).
It uses a IPC and TCP-socket to communicate with the daemon and provides a list of highlevel promise based methods.

Make sure to take a look at the [examples](https://github.com/RomanBurunkov/tm-node-mpd/tree/master/examples).

Note that 0.2 branch was experimental. Do not use is it at all.

Available options
------

You can pass options object to the MPD constructor.
The following options are available:

* **type** connection type('ipc' or 'network'), default value is 'network'.
* **ipc** path to the unix socket(ipc), default value is '/var/run/mpd/socket'.
* **host** mpd service host, default value is 'localhost'.
* **port** mpd service TCP port, default value is 6600.

Events
------

* ready - Emits after connected to the MPD service.
* error - Emits when any errors occured.
* update - Emits when got any updates from MPD service.
* status - Emits when MPD service status has been updated.
* disconnected - Emits after disconnecting from MPD service.

Methods
------

This part of doc is still in progress.
Please see the source code for all public methods.

* `connect` connects to the mpd.
* `disconnect` disconnects from the mpd.
* `alive` returns true if instance connected to the mpd.
* `command` sends a basic command specified in the first argument. Other arguments will be used as a commant parameters. E.g.

```javascript
  mpd.command('setvol', vol);
```

Example
------
This is a minimal exmaple which connects to a mpd running on localhost on the default port and prints the current playlist:

	const MPD = require('tm-node-mpd');
	const mpd = new MPD(); // pass { type: 'ipc' } to connect local mpd via unix socket.

```javascript
	mpd.on('ready', () => {
		for (let num = 0; num < mpd.playlist.length; num += 1) {
			const n = num + 1;
			console.log(n + ": " + mpd.playlist[num].artist + " - " + mpd.playlist[num].title);
		}
		mpd.disconnect();
	});
```

Make sure to take a look at the [examples](https://github.com/RomanBurunkov/tm-node-mpd/tree/master/examples).

License
-------
Copyright 2015 Frederick Gnodtke
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
# 0.3.3 (2021-03-07)

  * Fix: Protocol.parseKvp Add check that argument is a string.
  * Fix: Protocol.parseGreeting Add check that argument is a string.
  * Add support for Node 10(get rid of String.MatchAll issue #2).
  * Update github actions CI workflow.

# 0.3.2 (2021-02-20)

  * initialize a history log.
  * Move findReturn to proto section.
  * Add tests for return patterns.
  * Fix: Parse kvp when value contains whitespaces


tests/mpd.test.js
const Mpd = require('../src/mpd');

test('Mpd class exist', () => expect(Mpd).toBeDefined());

/* eslint-disable security-node/detect-crlf */
const mpd = new Mpd({ type: 'ipc' });
describe('Mpd short operation to change the volume.', () => {
  beforeAll(async () => {
    mpd.on('update', (changed) => {
      console.log(`Mpd on update: ${changed}`);
    });
    await mpd.connect();
    await new Promise((resolve) => setTimeout(resolve, 1000));
    await mpd.play();
  });
  afterAll(async () => {
    await new Promise((resolve) => setTimeout(resolve, 1000));
    await mpd.disconnect();
  });
  test('set volume to "60%"', async () => {
    expect.assertions(1);
    await expect(mpd.volume(60)).resolves.toBeUndefined();
  });
  test('volume should be "60%"', async () => {
    expect.assertions(1);
    await new Promise((resolve) => setTimeout(resolve, 3000));
    await expect(mpd.updateStatus()).resolves.toHaveProperty('volume', 0.6);
  });
  test('set volume to "100%"', async () => {
    expect.assertions(1);
    await expect(mpd.volume(100)).resolves.toBeUndefined();
  });
  test('volume should be "100%"', async () => {
    expect.assertions(1);
    await new Promise((resolve) => setTimeout(resolve, 3000));
    await expect(mpd.updateStatus()).resolves.toHaveProperty('volume', 1);
  });
});
/* eslint-enable security-node/detect-crlf */

describe('Mpd parse messages from "change events" while idle-ing.', () => {
  test('single: options', () => {
    expect.assertions(1);
    expect(
      mpd._matchAllChanged(`changed: options
OK`)
    ).toMatchObject(['options']);
  });
  test('multiple: mixer x3', () => {
    expect.assertions(1);
    expect(
      mpd._matchAllChanged(`changed: mixer
changed: mixer
changed: mixer
OK`)
    ).toMatchObject(['mixer', 'mixer', 'mixer']);
  });
  test('No matching event', () => {
    expect.assertions(1);
    expect(mpd._matchAllChanged('OK')).toHaveLength(0);
  });
  test('unexpected or void message', () => {
    expect.assertions(4);
    expect(mpd._matchAllChanged(undefined)).toHaveLength(0);
    expect(mpd._matchAllChanged(null)).toHaveLength(0);
    expect(mpd._matchAllChanged('')).toHaveLength(0);
    expect(mpd._matchAllChanged(9999)).toHaveLength(0);
  });
});


tests/song.test.js
const Song = require('../src/song');

const mockInfo = {
  time: 60,
  date: 'date',
  file: 'song.mp3',
  title: 'song title',
  track: 'song track',
  genre: 'song genre',
  artist: 'song artist',
  lastModified: 'lastModified',
};

test('Song class exist', () => expect(Song).toBeDefined());

test('Test flatCopy', () => {
  const song = new Song(mockInfo);
  expect(song.flatCopy).toBeDefined();
  expect(song.flatCopy()).toEqual(mockInfo);
});


tests/protocol.test.js
const proto = require('../src/protocol');

describe('Test parseKvp function', () => {
  test('Is defined', () => expect(proto.parseKvp).toBeDefined());
  test('Returns false if no data passed', () => expect(proto.parseKvp()).toBe(false));
  test('Returns false if passed string doesnt match kvp format', () => {
    expect(proto.parseKvp('no kvp')).toBe(false);
  });
  test('Returns false if passed argument is not a string', () => {
    [null, {}, 25, undefined].forEach((arg) => {
      expect(proto.parseKvp(arg)).toBe(false);
    });
  });
  test('Doesnt returns false if passed string matches kvp format', () => {
    expect(proto.parseKvp('vol: 42')).toBeTruthy();
  });
  test('Returned value has key prop if passed string matches kvp format', () => {
    const kvp = proto.parseKvp('vol: 42');
    expect(kvp.key).toBeDefined();
  });
  test('Returned value has val prop if passed string matches kvp format', () => {
    const kvp = proto.parseKvp('vol: 42');
    expect(kvp.val).toBeDefined();
  });
  test('Returns a kvp object if passed string matches kvp format', () => {
    const kvp = proto.parseKvp('vol: 42');
    expect(kvp.key).toBe('vol');
    expect(kvp.val).toBe('42');
  });
  test('Returnse values contain white spaces.', () => {
    expect.assertions(2);
    expect(proto.parseKvp('Artist: Various Artists')).not.toHaveProperty('val', 'Various');
    expect(proto.parseKvp('Title: Lazy fox jumps.')).toHaveProperty('val', 'Lazy fox jumps.');
  });
});

describe('Test parseGreeting function', () => {
  test('Is defined', () => expect(proto.parseGreeting).toBeDefined());
  test('Returns false if no data passed', () => expect(proto.parseGreeting()).toBe(false));
  test('Returns false if passed string doesnt match greetings format', () => {
    expect(proto.parseGreeting('Failed greetings')).toBe(false);
  });
  test('Returns false if passed argument is not a string', () => {
    [null, {}, 25, undefined].forEach((arg) => {
      expect(proto.parseGreeting(arg)).toBe(false);
    });
  });
  test('Doesnt returns false if passed string matches greetings format', () => {
    expect(proto.parseGreeting('OK MPD 0.20.2')).toBeTruthy();
  });
  test('Returned value has name prop if passed string matches greetings format', () => {
    const greeting = proto.parseGreeting('OK MPD 0.20.2');
    expect(greeting.name).toBeDefined();
  });
  test('Returned value has version prop if passed string matches greetings format', () => {
    const greeting = proto.parseGreeting('OK MPD 0.20.2');
    expect(greeting.version).toBeDefined();
  });
  test('Returns a protocol info object if passed string matches greetings format', () => {
    const greeting = proto.parseGreeting('OK MPD 0.20.2');
    expect(greeting.name).toBe('MPD');
    expect(greeting.version).toBe('0.20.2');
  });
});

describe('Test returnPatterns function', () => {
  test('Is defined', () => expect(proto.returnPatterns).toBeDefined());
  test('Returns an array', () => expect(Array.isArray(proto.returnPatterns())).toBe(true));
  test('Returned array is not empty', () => expect(proto.returnPatterns().length).toBeTruthy());
});

describe('Test findReturn function', () => {
  test('Is defined', () => expect(proto.findReturn).toBeDefined());
  test('Returns false if no data passed', () => expect(proto.findReturn()).toBe(false));
  test('Returns false if passed string doesn\'t contain mpd return mark', () => {
    expect(proto.findReturn('Some test message')).toBe(false);
  });
});


src/song.js
const { parseKvp } = require('./protocol');

const RES_OK = 'OK';
const ERR_MSG_UNKNOWN = 'Unknown response while parsing song.';

const FIELD_MAP = [
  { key: 'file', val: 'file' },
  { key: 'time', val: 'Time' },
  { key: 'date', val: 'Date' },
  { key: 'genre', val: 'Genre' },
  { key: 'title', val: 'Title' },
  { key: 'album', val: 'Album' },
  { key: 'track', val: 'Track' },
  { key: 'artist', val: 'Artist' },
  { key: 'lastModified', val: 'Last-Modified' },
];

module.exports = class Song {
  constructor(data) {
    const info = this._parseInfo(data);
    for (let key in info) {
      this[key] = info[key];
    }
  }

  flatCopy() {
    const obj = {};
    for (let key in this) {
      if (this.__proto__[key] !== undefined) continue;
      obj[key] = this[key];
    }
    return obj;
  }

  _parseInfo(data) {
    if (!Array.isArray(data)) return data;
    const info = {};
    data
      .filter((itm) => itm !== RES_OK)
      .forEach((itm) => {
        const kvp = parseKvp(itm);
        if (!kvp) throw new Error(ERR_MSG_UNKNOWN);
        const fieldInfo = FIELD_MAP.find((fldKvp) => fldKvp.val === kvp.key);
        if (!fieldInfo) return;
        info[fieldInfo.key] = kvp.val;
      });
    return info;
  }
};


src/protocol.js

/**
 * Parses KVP from MPD responses.
 * Some commands return data before the response ends with OK.
 * Each line is usually in the form NAME: VALUE.
 * @param {string} kvp String to parse for kvp.
 * @returns {Object|false} KVP object or false if parsing failed.
 */
module.exports.parseKvp = (kvp = '') => {
  if (!kvp || typeof kvp !== 'string') {
    return false;
  }
  const m = kvp.match(/(\S+)\s*:\s*(.+)$/);
  return Array.isArray(m) && m.length === 3
    ? { key: m[1].trim(), val: m[2].trim() }
    : false;
};

/**
 * Parses MPD Greeting message
 * According to the MPD protocol documentation when the client connects to the server,
 * the server will answer with the following line: 'OK MPD version'
 * where version is a protocol version identifier such as 0.12.2.
 * @param {string} message MPD greeting message.
 * @returns {Object|false} mpd proto details: { name: service name(MPD), version: proto version}.
 */
module.exports.parseGreeting = (message = '') => {
  if (!message || typeof message !== 'string') {
    return false;
  }
  const m = message.match(/OK\s(.+)\s(.+)/);
  return Array.isArray(m) && m.length === 3
    ? { name: m[1], version: m[2] }
    : false;
};

/**
 * MPD protocol has several return patterns.
 * @returns {Array} Array with supported mpd return patterns.
 */
module.exports.returnPatterns = () => [
  /OK(?:\n|$)/g,
  /ACK\s*\[\d*\@\d*]\s*\{.*?\}\s*.*?(?:$|\n)/g
];

/**
 * Searches for an mpd protocol return mark in the response data.
 * @param {string} message MPD message.
 * @returns {number|false} Total message length or false if no marks has been found.
 */
module.exports.findReturn = (message = '') => {
  if (!message) return false;
  for (let pattern of module.exports.returnPatterns()) {
    const arr = pattern.exec(message);
    if (arr) return arr.index + arr[0].length;
  }
  return false;
};

/**
 * Parses changes/updates in mpd idle messages.
 * @param {string} message MPD message.
 * @returns {Array} List of found changes.
 */
module.exports.parseChanged = (message = '') => {
  if (!message || typeof message !== 'string') {
    return [];
  }
  return message.split('\n')
    .filter(l => l.indexOf('changed:') === 0)
    .map(l => l.substring(8).trim());
};


examples/play.js
const MPD = require('../');

const settings = {
  host: process.env.MPD_HOST || 'localhost',
  port: process.env.MPD_PORT || 6600
};

console.log(`Creating MPD instance for service on ${settings.host}:${settings.port}`);
const mpd = new MPD(settings);

mpd.on('ready', async () => {
  try {
    console.log(`MPD connection is ready, protocol ver ${mpd.server.version}.`)
    console.log(mpd.status);
    if (mpd.songs.length) console.log(mpd.songs);
    if (mpd.playlist.length) console.log(mpd.playlist);
    console.log(`Sending play command...`);
    await mpd.play();
    console.log(`Command play has been successfully sent.`);
  } catch (e) {
    console.error(e);
  }
});

mpd.on('update', (status)  => {
  console.log('Update:', status);
  switch(status) {
    case 'mixer':
    case 'player':
    case 'options': return console.log('Status after update', mpd.status);
    case 'playlist': return console.log('Playlist after update', mpd.playlist);
    case 'database': return console.log('Songs after update', mpd.songs);
  }
});

mpd.connect();


examples/volume.js
const MPD = require('../');

const volume = process.argv.length > 2 ? parseInt(process.argv[2], 10) : 30;
const settings = {
  host: process.env.MPD_HOST || 'localhost',
  port: process.env.MPD_PORT || 6600
};

console.log(`Creating MPD instance for service on ${settings.host}:${settings.port}`);
const mpd = new MPD(settings);

mpd.on('ready', async () => {
  try {
    console.log(`MPD connection is ready, protocol ver ${mpd.server.version}.`)
    console.log(`Sending volume ${volume} command...`);
    await mpd.volume(volume);
    console.log(`Command volume ${volume} has been successfully sent.`);
  } catch (e) {
    console.error(e);
  } finally {
    console.log(`Disconnecting from ${settings.host}:${settings.port}`);
    mpd.disconnect();
  }
});

console.log(`Connecting to the MPD service...`);
mpd.connect();


examples/devtests.js
const MPD = require('../');

const settings = {
  host: process.env.MPD_HOST || 'localhost',
  port: process.env.MPD_PORT || 6600
};

console.log(`Creating MPD instance for service on ${settings.host}:${settings.port}`);
const mpd = new MPD(settings);

// mpd.on('idle', (idle, data) => console.log('Idle: ', idle, data));
mpd.on('command', (data) => console.log('Sending command: ', data));
mpd.on('error', e => console.error(e));

mpd.on('ready', async () => {
  try {
    console.log(`MPD connection is ready, protocol ver ${mpd.server.version}.`)

    console.log(`Adding mock songs...`);
    const songs = [
      'Nobara_Hayakawa_-_01_-_Trail.mp3',
      'Monopole_-_02_-_Stereo-vision_radio.mp3',
      'Robin_Grey_-_01_-_These_Days.mp3'
    ]
    for (const song of songs) {
      await mpd.volume(25);
      console.log(`Command volume has been successfully sent.`)
      await mpd.add(song);
      console.log(`Command add has been successfully sent.`)
    }

    // console.log(mpd.status);
    // if (mpd.songs.length) console.log(mpd.songs);
    // if (mpd.playlist.length) console.log(mpd.playlist);
    setTimeout(async () => {
      await mpd.play();
      console.log(`Command play has been successfully sent.`);
    }, 1000);


    setTimeout(async () => {
      await mpd.stop();
      console.log(`Command stop has been successfully sent.`);
    }, 5000);
  } catch (e) {
    console.error(e);
  }
});

mpd.on('update', (status)  => {
  console.log('Update:', status);
  // switch(status) {
  //   case 'mixer':
  //   case 'player':
  //   case 'options': return console.log('Status after update', mpd.status);
  //   case 'playlist': return console.log('Playlist after update', mpd.playlist);
  //   case 'database': return console.log('Songs after update', mpd.songs);
  // }
});

mpd.connect();


src/mpd.js
const Song = require('./song');
const { Socket } = require('net');
const { EventEmitter } = require('events');
const { parseKvp, parseGreeting, findReturn } = require('./protocol');

const DEF_PORT = 6600;
const DEF_HOST = 'localhost';
const DEF_SOCKET = '/var/run/mpd/socket';
const DEF_KEEP_ALIVE = false;
const DEF_CONN_TYPE = 'network';
const CONN_TYPES = ['ipc', 'network'];
const RECONNECT_INTERVAL = 5000;
const CONST_FILE_LINE_START = 'file:';
const GENERIC_COMMANDS = ['play', 'stop', 'pause', 'next', 'previous', 'toggle', 'clear'];

const buffer = Symbol('Read buffer');

module.exports = class MPD extends EventEmitter {
  /**
   * MPD connection constructor.
   * @param {Object} options MPD options.
   * @param {string} options.ipc Path to the IPC(Unix Domain Socket).
   * @param {string} options.host MPD service host.
   * @param {number} options.port MPD service TCP port.
   * @param {string} options.type MPD connection type: ipc/network.
   * @param {boolean} options.keepAlive Use keep alive for MPD network connection.
   */
  constructor(options = {}) {
    super();
    // Applying options.
    this.ipc = options.ipc || DEF_SOCKET;
    this.host = options.host || DEF_HOST;
    this.port = options.port || DEF_PORT;
    this.type = CONN_TYPES.includes(options.type) ? options.type : DEF_CONN_TYPE;
    this.keepAlive = !!options.keepAlive || DEF_KEEP_ALIVE;
    // Init props.
    this.songs = [];
    this.status = {};
    this.server = {};
    this[buffer] = '';
    this.playlist = [];
    this._requests = [];
    this.connected = false;
    this.disconnecting = false;
    this._initGenericCommand();
    this.on('disconnected', () => this.restoreConnection());
    return this;
  }

  /**
   * Sends a simple command specified in arguments to the mpd instance.
   * First argument should be a command name.
   * All further args will be uses as a command parameters.
   * @returns {Promise}
   */
  async command() {
    const r = await this._sendCommand(...arguments);
    return this._answerCallbackError(r);
  }

  async query() {
    const r = await this._sendCommand(...arguments);
    const arr = r.split('\n');
    const res_status = arr.pop();
    await this._answerCallbackError(res_status);
    const data = {};
    try {
      for (const line of arr) {
        const kvp = parseKvp(line);
        data[kvp.key] = kvp.val;
      }
    } catch (e) {
      throw new Error(`An error occurred while parsing the query response. ${e}`);
    }
    return data;
  }

  alive() {
    return this.connected;
  }

  async add(name) {
    return this.command('add', name);
  }

  async playId(id) {
    return this.command('play', id);
  }

  async deleteId(id) {
    return this.command('delete', id);
  }

  async volume(vol) {
    return this.command('setvol', vol);
  }

  async repeat(repeat = 1) {
    return this.command('repeat', repeat);
  }

  async crossfade(seconds = 0) {
    return this.command('crossfade', seconds);
  }

  async seek(songId, time) {
    return this.command('seek', songId, time);
  }

  async updateSongs() {
    const r = await this._sendCommand('update');
    const arr = r.split(/\n/);
    return this._answerCallbackError(arr[1]);
  }

  async searchAdd(search) {
    let args = ['searchadd'];
    for (let key in search) {
      args.push(key);
      args.push(search[key]);
    }
    return this.command(...args);
  }

  /**
   * Connect and disconnect
   */
  connect() {
    try {
      this.client = new Socket();
      this.client.setEncoding('utf8');
      this.connected = false;
      this.commanding = true;
      this.disconnecting = false;
      this.client.once('end', () => {
        if (this.disconnecting) return;
        this.connected = false;
        this.emit('disconnected');
      });
      this.client.on('error', (e) => {
        this.connected = false;
        this.emit('error', e);
        this.emit('disconnected');
      });
      this.client.on('connect', () => {
        this.connected = true;
        clearInterval(this.reconnectInterval);
        this.reconnectInterval = null;
        this.client.once('data', async (data) => {
          await this._initialGreeting(data);
        });
      });
      // Connecting to the MPD via IPC or TCP.
      this.client.connect(...(this.type === 'ipc' ? [this.ipc] : [this.port, this.host]));
    } catch (e) {
      this.restoreConnection();
    }
  }

  restoreConnection() {
    if (this.reconnectInterval) {
      clearInterval(this.reconnectInterval);
    }
    this.reconnectInterval = setInterval(() => {
      this.disconnect();
      this.connect();
    }, RECONNECT_INTERVAL);
  }

  disconnect() {
    this.disconnecting = true;
    this.busy = false;
    this._activeListener = null;
    this._requests.splice(0, this._requests.length);
    if (this.client) {
      this.client.destroy();
      delete this.client;
    }
  }

  /**
   * Not-so-toplevel methods
   */

  _setReady() {
    this.emit('ready', this.status, this.server);
  }

  async _answerCallbackError(msg) {
    if (msg === 'OK') return;
    throw new Error(`Bad status: "${msg}" after command "${this._activeMessage}"`);
  }

  async _initGenericCommand() {
    for (let cmd of GENERIC_COMMANDS) {
      this[cmd] = await this.command.bind(this, [cmd]);
    }
  }

  async _updatePlaylist() {
    const message = await this._sendCommand('playlistinfo');
    const lines = message.split('\n');
    this.playlist = [];
    let songLines = [];
    let pos;
    for (let i = 0; i < lines.length - 1; i += 1) {
      let line = lines[i];
      if (i !== 0 && line.startsWith(CONST_FILE_LINE_START)) {
        this.playlist[pos] = new Song(songLines);
        songLines = [];
        pos = -1;
      }
      if (line.startsWith('Pos')) {
        pos = parseInt(line.split(':')[1].trim(), 10);
      } else {
        songLines.push(line);
      }
    }
    if (songLines.length !== 0 && pos !== -1) {
      this.playlist[pos] = new Song(songLines);
    }
    await this._answerCallbackError(lines[lines.length - 1]);
    return this.playlist;
  }

  async _updateSongs() {
    const message = await this._sendCommand('listallinfo');
    const lines = message.split('\n');
    this.songs = [];
    let songLines = [];
    for (let i = 0; i < lines.length - 1; i += 1) {
      let line = lines[i];
      if (i !== 0 && line.startsWith(CONST_FILE_LINE_START)) {
        this.songs.push(new Song(songLines));
        songLines = [];
      }
      songLines.push(line);
    }
    if (songLines.length !== 0) {
      this.songs.push(new Song(songLines));
    }
    await this._answerCallbackError(lines[lines.length - 1]);
    return this.songs;
  }

  _parseStatusResponseValue({ key, val }) {
    switch (key) {
      case 'repeat':
      case 'single':
      case 'random':
      case 'consume':
        return val === '1';
      case 'song':
      case 'xfade':
      case 'bitrate':
      case 'playlist':
      case 'playlistlength':
        return parseInt(val, 10);
      case 'volume':
        return parseFloat(val.replace('%', '')) / 100;
      case 'time': {
        const times = val.split(':');
        return { elapsed: times[0], length: times[1] };
      }
      default:
        return val;
    }
  }

  async _parseStatusResponse(message) {
    try {
      for (const line of message.split('\n')) {
        if (line === 'OK') continue;
        const kvp = parseKvp(line);
        this.status[kvp.key] = this._parseStatusResponseValue(kvp);
      }
    } catch (e) {
      throw new Error(`Unknown response while fetching status: ${e}`);
    }
    return this.status;
  }

  async updateStatus() {
    const r = await this._sendCommand('status');
    return this._parseStatusResponse(r);
  }

  /*
   * Message handling
   */

  /**
   * Initiate MPD connection with greeting message.
   * @param {string} message
   */
  async _initialGreeting(message) {
    try {
      this.server = parseGreeting(message);
    } catch (e) {
      this.restoreConnection();
      throw new Error(`Unexpected greeting message: '${message}'! ${e}`);
    }
    if (this.type === 'network' && this.keepAlive) {
      this.client.setKeepAlive(this.keepAlive);
    }
    this.client.on('data', async (d) => {
      await this._onData(d);
    });
    try {
      await this.updateStatus();
      await this._updateSongs();
      await this._updatePlaylist();
      await this._setReady();
    } catch (e) {
      this.emit('error', e);
    }
  }

  async _onData(data) {
    if (!this.idling && !this.commanding) return;
    this[buffer] += !data ? '' : data.trim();
    let index;
    try {
      index = findReturn(this[buffer]);
    } catch (e) {
      return;
    }
    // We found a return mark
    const string = this[buffer].substring(0, index).trim();
    this[buffer] = this[buffer].substring(index, this[buffer].length);
    if (this.idling) {
      await this._onMessage(string);
    } else if (this.commanding) {
      await this._handleResponse(string);
    }
  }

  /**
   * Idling
   */

  _checkIdle() {
    if (this._activeListener || this._requests.length || this.idling) return;
    this._enterIdle();
  }

  _enterIdle() {
    this.idling = true;
    this.commanding = false;
    this._write('idle');
  }

  _leaveIdle(callback) {
    this.client.once('data', () => {
      this.idling = false;
      this.commanding = true;
      callback();
    });
    this._write('noidle');
  }

  /**
   * Handle idle mode updates.
   * @param {string} message Message from MPD.
   */
  async _onMessage(message) {
    try {
      this.idling = false;
      this.commanding = true;
      // It is possible to get a change event or just OK message
      if (message.match(/^\s*OK/)) return;
      const matches = this._matchAllChanged(message);
      if (!matches.length) {
        this.restoreConnection();
        throw new Error(`Received unknown message during idle: ${message}`);
      }
      for (const update of matches) {
        const afterUpdate = () => {
          this.emit('update', update);
          this.emit('status', update);
        };
        switch (update) {
          case 'mixer':
          case 'player':
          case 'options':
            await this.updateStatus();
            afterUpdate();
            break;
          case 'playlist':
            await this._updatePlaylist();
            afterUpdate();
            break;
          case 'database':
            await this._updateSongs();
            afterUpdate();
            break;
        }
      }
      this._checkIdle();
    } catch (e) {
      this.emit('error', e);
    }
  }

  _matchAllChanged(message = '') {
    const data = [];
    if (typeof message !== 'string' || message === '') {
      return data;
    }
    for (const line of message.split('\n')) {
      if (line.indexOf('changed:') === 0) {
        data.push(line.substring(8).trim());
      }
    }
    return data;
  }

  /**
   * Sending messages
   */

  _dequeue(request) {
    this.busy = false;
    this._activeListener = request.callback;
    this._activeMessage = request.message;
    this._write(request.message);
  }

  _checkOutgoing() {
    if (this._activeListener || this.busy) return;
    const request = this._requests.shift();
    if (!request) return;
    this.busy = true;
    return this.idling ? this._leaveIdle(() => this._dequeue(request)) : this._dequeue(request);
  }

  async _sendCommand() {
    if (arguments.length === 0) return;
    const cmd = arguments[0];
    let args = '';
    for (let i = 1; i < arguments.length; i += 1) {
      args += ' "' + arguments[i] + '" ';
    }
    return this._send(cmd + args);
  }

  async _send(message) {
    return new Promise((resolve, reject) => {
      try {
        this._requests.push({ message, callback: resolve, errorback: reject });
        this._checkOutgoing();
      } catch (e) {
        reject(e);
      }
    });
  }

  async _handleResponse(message) {
    if (!this._activeListener) return;
    const callback = this._activeListener;
    this._activeListener = null;
    this._checkOutgoing();
    this._checkIdle();
    await callback(message);
    return;
  }

  _write(text) {
    try {
      if (!this.connected) {
        this.restoreConnection();
        throw new Error('Disconnect while writing to MPD: ' + text);
      }
      this.client.write(text + '\n');
    } catch (e) {
      this.emit('error', e);
    }
  }
};
